kivy/core/text/_text_pango.pyx

```diff
@@ -0,0 +1,372 @@
+    cc.metrics_lang = pango_language_get_default()
+    cc.metrics = pango_context_get_metrics(cc.context, cc.fontdesc,
+                                           pango_language_get_default())
+    if not cc.metrics:
+        Logger.warn("_text_pango: Could not get context metrics")
+        cc.__dealloc__()
+        return
+
+    # Create layout from context
+    cc.layout = pango_layout_new(cc.context)
+    if not cc.layout:
+        Logger.warn("_text_pango: Could not create pango layout")
+        cc.__dealloc__()
+        return
+
+    # If autodir is false, the context's base direction is used (set above)
+    if text_dir == PANGO_DIRECTION_NEUTRAL:
+        pango_layout_set_auto_dir(cc.layout, TRUE)
+    else:
+        pango_layout_set_auto_dir(cc.layout, FALSE)
+
+    # The actual font size is specified in pango markup, and the
+    # actual font is whatever TTF is loaded in this context. This
+    # may not be needed at all.
+    cc.fontdesc = pango_font_description_from_string("Arial")
+    pango_layout_set_font_description(cc.layout, cc.fontdesc)
+
+    # FIXME: does this need to change w/label settings?
+    #pango_layout_set_alignment(cc.layout, PANGO_ALIGN_LEFT)
+    #pango_layout_set_spacing(cc.layout, n)
+
+    _add_pango_cache(fontid, cc)
+    return cc
+
+
+# Renders the pango layout to a grayscale bitmap, and blits RGBA at x, y
+@cython.cdivision(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+cdef _render_context(ContextContainer cc, unsigned char *dstbuf,
+                     int x, int y, int final_w, int final_h,
+                     unsigned char textcolor[]):
+    if not dstbuf or final_w <= 0 or final_h <= 0 or x > final_w or y > final_h:
+        Logger.warn('_text_pango: Invalid blit: final={}x{} x={} y={}'
+                    .format(final_w, final_h, x, y))
+        return
+
+    # Note, w/h refers to the current subimage size, final_w/h is end result
+    cdef int w, h
+    pango_layout_get_pixel_size(cc.layout, &w, &h)
+    if w <= 0 or h <= 0 or x + w > final_w or y + h > final_h:
+        Logger.warn('_text_pango: Invalid blit: final={}x{} x={} y={} w={} h={}'
+                    .format(final_w, final_h, x, y, w, h))
+        return
+
+    cdef FT_Bitmap bitmap
+    cdef int xi, yi
+    cdef unsigned char graysrc
+    cdef unsigned long grayidx
+    cdef unsigned long yi_w
+    cdef unsigned long offset
+    cdef unsigned long offset_fixed = x + (y * final_w)
+    cdef unsigned long offset_yi = final_w - w
+    cdef unsigned long maxpos = final_w * final_h
+
+    # Sanity check that we don't go out of bounds here, should not happen
+    if offset_fixed + ((h-1)*offset_yi) + ((h-1)*w) + w - 1 > maxpos:
+        Logger.warn('_text_pango: Ignoring out of bounds blit: final={}x{} '
+                    'x={} y={} w={} h={} maxpos={}'.format(
+                    final_w, final_h, x, y, w, h, maxpos))
+        return
+
+    with nogil:
```

Comment:
I wonder the impact of nogil at this point. Just a reflexion, i know we do it from times to times, but i'm thinking that by the time we need to acquire, would we wait more than just keep it?

Reply:
Maybe you could put the whole function nogil, but then no Logger